rust/rope/src/tree.rs

```diff
@@ -116,29 +123,48 @@ enum NodeVal<N: NodeInfo> {
 
 // also consider making Metric a newtype for usize, so type system can
 // help separate metrics
+
+/// A trait for quickly processing attributes of a NodeInfo.
+/// 
+/// For the conceptual background see the 
+/// [blog post, Rope science, part 2: metrics](https://github.com/google/xi-editor/blob/master/doc/rope_science/rope_science_02.md).
 pub trait Metric<N: NodeInfo> {
+    /// Return the number of boundarys in the NodeInfo::Leaf
+    /// 
+    /// TODO: What does the second argument do?
     fn measure(&N, usize) -> usize;
 
+    /// Return a offset in base units forwich self.from_base_units will return the argument.
+    /// 
+    /// TODO: Does it have to be the largest, smallest, or just eny offset?
     fn to_base_units(l: &N::L, in_measured_units: usize) -> usize;
 
+    /// Return the number of boundarys in the NodeInfo::Leaf befor the offset in_base_units.
     fn from_base_units(l: &N::L, in_base_units: usize) -> usize;
 
-    // The next three methods work in base units.
-
-    // These methods must indicate a boundary at the end of a leaf,
-    // if present. A boundary at the beginning of a leaf is optional
-    // (the previous leaf will be queried).
-
+    /// Return whether the offset in base units is a boundary of this metric.
+    /// If a boundary is at end of a leaf then this method must return true.
+    /// However, A boundary at the beginning of a leaf is optional
+    /// (the previous leaf will be queried).
     fn is_boundary(l: &N::L, offset: usize) -> bool;
 
-    // will be called with offset > 0
+    /// Return the largest usize smaller then offset forwich is_boundary will return true.
+    /// will be called with offset > 0.
+    /// 
+    /// TODO: If offset is a boundary can this return offset, or does it have to return a strictly smaller number?
     fn prev(l: &N::L, offset: usize) -> Option<usize>;
 
+    /// Return the smallest usize larger then offset forwich is_boundary will return true.
+    /// will be called with offset > 0.
+    /// 
+    /// TODO: If offset is a boundary can this return offset, or does it have to return a strictly smaller number?
     fn next(l: &N::L, offset: usize) -> Option<usize>;
 
-    // When can_fragment is false, the ends of leaves are always
-    // considered to be boundaries. More formally:
-    // !can_fragment -> to_base_units(measure) = leaf.len
+    /// When can_fragment is false, the ends of leaves are always
+    /// considered to be boundaries. More formally:
+    /// !can_fragment -> to_base_units(measure) = leaf.len
+    /// 
+    /// TODO: What dose this mean?
```

Comment:
"Returns `true` if the measured units in this metric can span multiple leaves.  As an example, in a metric that measures lines in a rope, a line may start in one leaf and end in another; however in a metric measuring bytes, storage of a single byte cannot extend across leaves."

Reply:
Ok, but I don't understand:
 - The code fragment in the doc.
 - Why and by whom this would be called.

Reply:
I don't know much better than you; I'm mostly just grepping through the code and taking my best guess, so apologies if my explanations aren't the best.

The little shorthand statement in the doc is saying, "'can_fragment() == false' _implies_  the number of base units in this leaf is equal to `leaf.len()`". (Raph is using logical notation; I assume `->` is the [material conditional](https://en.wikipedia.org/wiki/Material_conditional) operator.)

I'm not sure if there's a case where this would be used by a general library consumer; it is [used internally](https://github.com/google/xi-editor/blob/master/rust/rope/src/tree.rs#L521) when doing boundary calculations.

Reply:
This is used in `Tree::convert_metrics` as an optimization. The motivating example is definitely counting line ends, which can fragment. Looking at the code, I'm actually not 100% sure how important the optimization is. I _think_ this might be equivalent to saying that it's possible for a leaf to have zero measure.

Reply:
Thinking about it more, I think `can_fragment` is "not atomic" in the terminology of #456. I think it's useful to preserve, but maybe "atomic" (and inverting the boolean) is clearer. I agree, the way I wrote the invariant is confusing.